src/radical/pilot/agent/agent_0.py

```diff
@@ -340,77 +353,130 @@ def _start_services(self):
+            #                         'mem'      : 0}]}
+        # }
+
+        # launcher = self._rm.find_launcher(service_task)
+        # if not launcher:
+        #     raise RuntimeError('no launch method found for sub agent')
 
-        tmp  = '#!/bin/sh\n\n'
-        tmp += 'export RP_PILOT_SANDBOX="%s"\n\n' % self._pwd
-        cmds = launcher.get_launcher_env()
-        for cmd in cmds:
-            tmp += '%s || exit 1\n' % cmd
-        cmds = launcher.get_launch_cmd(service_task, ex_name)
-        tmp += '%s\nexit $?\n\n' % '\n'.join(cmds)
+        # FIXME: set RP environment (as in Popen Executor)
 
-        with ru.ru_open(ls_name, 'w') as fout:
-            fout.write(tmp)
+        # tmp  = '#!/bin/sh\n\n'
+        # tmp += 'export RP_PILOT_SANDBOX="%s"\n\n' % self._pwd
+        # cmds = launcher.get_launcher_env()
+        # for cmd in cmds:
+        #     tmp += '%s || exit 1\n' % cmd
+        # cmds = launcher.get_launch_cmd(service_task, ex_name)
+        # tmp += '%s\nexit $?\n\n' % '\n'.join(cmds)
+        #
+        # with ru.ru_open(ls_name, 'w') as fout:
+        #     fout.write(tmp)
 
-        tmp  = '#!/bin/sh\n\n'
-        tmp += '. ./env/service.env\n'
-        tmp += '/bin/sh -l ./services\n\n'
+        # tmp  = '#!/bin/sh\n\n'
+        # tmp += '. ./env/service.env\n'
+        # tmp += '/bin/sh -l ./services\n\n'
+        #
+        # with ru.ru_open(ex_name, 'w') as fout:
+        #     fout.write(tmp)
+        #
+        #
+        # make sure scripts are executable
+        # st_l = os.stat(ls_name)
+        # st_e = os.stat(ex_name)
+        # os.chmod(ls_name, st_l.st_mode | stat.S_IEXEC)
+        # os.chmod(ex_name, st_e.st_mode | stat.S_IEXEC)
+        #
+        # spawn the sub-agent
+        # cmdline = './%s' % ls_name
+        #
+        # self._log.info('create services: %s' % cmdline)
+        # ru.sh_callout_bg(cmdline, stdout='services.out', stderr='services.err')
 
-        with ru.ru_open(ex_name, 'w') as fout:
-            fout.write(tmp)
+        # self.publish(rpc.STATE_PUBSUB, {'cmd': 'service_state_update','arg': services})
 
+    def _state_cb_of_services(self, topic, msg):
+        cmd = msg['cmd']
+        services = msg['arg']
 
-        # make sure scripts are executable
-        st_l = os.stat(ls_name)
-        st_e = os.stat(ex_name)
-        os.chmod(ls_name, st_l.st_mode | stat.S_IEXEC)
-        os.chmod(ex_name, st_e.st_mode | stat.S_IEXEC)
+        self._log.info('received services for launching: %s' % services)
 
-        # spawn the sub-agent
-        cmdline = './%s' % ls_name
+        self._log.debug('services started done')
 
-        self._log.info('create services: %s' % cmdline)
-        ru.sh_callout_bg(cmdline, stdout='services.out', stderr='services.err')
+        running_services = []
 
-        self._log.debug('services started done')
+        if cmd == 'update':
 
+            for service in ru.as_list(services):
+                if "service" in service['uid']:
```

Comment:
What happens if a user starts a task and assignes it the uid `service.foo.1`?  This code will go boom...

3 lines below you have a different check `if uid in self._service_task_ids` - that seems much safer.  You only need to check once really, so I would suggest to use the second one here.

Reply:
Please don't resolve comments which are not addressed.  Again: the policy is that the person who comments is also the one who resolves the comment.

Reply:
resolved.